pkg/controller/admissionchecks/preemption/preemption.go

```diff
@@ -0,0 +1,287 @@
+				if err != nil {
+					log.V(2).Error(err, "Unable to issue preemption")
+					c.trigger("retryEviction", klog.KObj(wl.Obj))
+				} else {
+					log.V(2).Info("Preemption triggered", "count", count)
+				}
+			}
+		}
+		// 3 add it back to the Snapshot
+		snapshot.AddWorkload(wl)
+	}
+}
+
+func filterWorkloads(snapsot *cache.Snapshot) []*workload.Info {
+	preemptNow := []*workload.Info{}
+	preemptLater := []*workload.Info{}
+	for _, cq := range snapsot.ClusterQueues {
+		now, later := cq.GetPreemptingWorklods(snapsot.AdmissionChecks)
+		preemptNow = append(preemptNow, now...)
+		preemptLater = append(preemptLater, later...)
+	}
+
+	// remove the workloads that should not evaluated now to limit the number of evictions
+	// needed at the current point in time
+	for _, wl := range preemptLater {
+		snapsot.RemoveWorkload(wl)
+	}
+	return preemptNow
+}
+
+func totalRequestsForWorkload(wl *workload.Info) cache.FlavorResourceQuantities {
+	usage := make(cache.FlavorResourceQuantities)
+	for _, ps := range wl.TotalRequests {
+		for res, q := range ps.Requests {
+			flv := ps.Flavors[res]
+			resUsage := usage[flv]
+			if resUsage == nil {
+				resUsage = make(map[corev1.ResourceName]int64)
+				usage[flv] = resUsage
+			}
+			resUsage[res] += q
+		}
+	}
+	return usage
+}
+
+func resourcesNeedingPreemption(wl *workload.Info, usage cache.FlavorResourceQuantities, snap *cache.Snapshot) preemption.ResourcesPerFlavor {
+	ret := make(preemption.ResourcesPerFlavor)
+
+	cq := snap.ClusterQueues[wl.ClusterQueue]
+	for _, rg := range cq.ResourceGroups {
+		for _, flvQuotas := range rg.Flavors {
+			flvReq, found := usage[flvQuotas.Name]
+			if !found {
+				// Workload doesn't request this flavor.
+				continue
+			}
+			cqResUsage := cq.Usage[flvQuotas.Name]
+			var cohortResUsage, cohortResRequestable map[corev1.ResourceName]int64
+			if cq.Cohort != nil {
+				cohortResUsage = cq.Cohort.Usage[flvQuotas.Name]
+				cohortResRequestable = cq.Cohort.RequestableResources[flvQuotas.Name]
+			}
+			for rName, rReq := range flvReq {
+				limit := flvQuotas.Resources[rName].Nominal
+				if flvQuotas.Resources[rName].BorrowingLimit != nil {
+					limit += *flvQuotas.Resources[rName].BorrowingLimit
+				}
+				exceedsNominal := cqResUsage[rName]+rReq > limit
+				exceedsBorrowing := cq.Cohort != nil && cohortResUsage[rName]+rReq > cohortResRequestable[rName]
+				if exceedsNominal || exceedsBorrowing {
+					if _, found := ret[flvQuotas.Name]; !found {
+						ret[flvQuotas.Name] = sets.New(rName)
+					} else {
+						ret[flvQuotas.Name].Insert(rName)
+					}
+				}
+			}
+		}
+	}
+	return ret
+}
+
+func updateCheckStatus(ctx context.Context, c client.Client, wl *kueue.Workload, successful bool) error {
+	cond := metav1.Condition{
+		Type:    constants.PreemptionAdmissionCheckName,
+		Status:  metav1.ConditionTrue,
+		Reason:  kueue.CheckStateReady, // the reason is not relevant, add this to keep it aligned wit the other checks
+		Message: "Preemption done",
```

Comment:
The code is a bit confusing, maybe consider:

if successful {

} else {


}

Reply:
done